### PR TITLE
 discovery via hazelcast.cloud is implemented for client

### PR DESCRIPTION
**Pull request by sancar**

### CODE REVIEW

**Comment by sancar** [comment]
run-lab-run

**Comment by sancar** [comment]
run-lab-run

**Comment by sancar** [comment]
run-lab-run

File: hazelcast-client/src/main/java/com/hazelcast/client/config/ClientCloudConfig.java [hunk 1]
```diff
@@ -0,0 +1,58 @@
...
+ */
+
+package com.hazelcast.client.config;
+
+/**
+ * hazelcast.cloud configuration to let the client connect the cluster via hazelcast.cloud
+ */
+public class ClientCloudConfig {
+
+    private String token;
```
Comment (enesakar) [inline hazelcast-client/src/main/java/com/hazelcast/client/config/ClientCloudConfig.java]: can we change `token` with `discoveryToken` everywhere it appears.
token created wrong assumptions related to security.

File: hazelcast-client/src/main/java/com/hazelcast/client/spi/properties/ClientProperty.java [hunk 2]
```diff
@@ -183,6 +183,12 @@
     public static final HazelcastProperty RESPONSE_THREAD_COUNT
             = new HazelcastProperty("hazelcast.client.response.thread.count", 2);
 
+
+    /**
+     * Token to use when discovering cluster via hazelcast.cloud
+     */
+    public static final HazelcastProperty HAZELCAST_CLOUD_TOKEN = new HazelcastProperty("hazelcast.client.cloud.token");
```
Comment (enesakar) [inline hazelcast-client/src/main/java/com/hazelcast/client/spi/properties/ClientProperty.java]: same here.
we need to use `hazelcast.client.cloud.discovery.token`

File: hazelcast-client/src/main/java/com/hazelcast/client/impl/HazelcastClientInstanceImpl.java [hunk 3]
```diff
@@ -289,18 +304,112 @@ private MetricsRegistryImpl initMetricsRegistry() {
...
+        if (awsDiscoveryEnabled) {
+            count++;
+        }
+        if (discoverySpiEnabled) {
+            count++;
+        }
+        if (hazelcastCloudEnabled) {
+            count++;
+        }
+        if (count > 1) {
```
Comment (ihsandemir) [inline hazelcast-client/src/main/java/com/hazelcast/client/impl/HazelcastClientInstanceImpl.java]: I believe that our discussion resulted that we may allow member address list together with the discovery, right?

File: hazelcast-client/src/test/java/com/hazelcast/client/spi/impl/discovery/ConflictingConfigTest.java [hunk 4]
```diff
@@ -0,0 +1,139 @@
...
+        config.getNetworkConfig().getCloudConfig().setEnabled(true);
+        hazelcastFactory.newHazelcastClient(config);
+    }
+
+
+    @Test(expected = IllegalStateException.class)
+    public void testClusterMembersGiven_and_HazelcastCloudViaProperty() {
+        ClientConfig config = new ClientConfig();
+        config.getNetworkConfig().addAddress("127.0.0.1");
+        config.setProperty(ClientProperty.HAZELCAST_CLOUD_TOKEN.getName(), "TOKEN");
```
Comment (ihsandemir) [inline hazelcast-client/src/test/java/com/hazelcast/client/spi/impl/discovery/ConflictingConfigTest.java]: see comment above (this is legal usage, should not throw)

File: hazelcast-client/src/main/resources/hazelcast-client-full.xml [hunk 5]
```diff
@@ -112,6 +112,9 @@
             <max-attempts>5</max-attempts>
             <echo-fail-fast-on-startup>false</echo-fail-fast-on-startup>
         </icmp-ping>
+        <hazelcast-cloud enabled="false">
+            <token>EXAMPLE_TOKEN</token>
```
Comment (sancar) [inline hazelcast-client/src/main/resources/hazelcast-client-full.xml]: Changing this to discovery-token as well @enesakar

**Comment by sancar** [comment]
run-lab-run

File: hazelcast-client/src/main/java/com/hazelcast/client/impl/HazelcastClientInstanceImpl.java (hunk 3 above)
Comment (sancar) [reply hazelcast-client/src/main/java/com/hazelcast/client/impl/HazelcastClientInstanceImpl.java]: It was vice verse. You can check the final decision from related Design Document

File: hazelcast-client/src/test/java/com/hazelcast/client/spi/impl/discovery/HazelcastCloudTranslatorTest.java [hunk 6]
```diff
@@ -0,0 +1,111 @@
...
+    @Test
+    public void testTranslate_whenAddressIsMapped_thenReturnPublicAddress() {
+        Address actual = translator.translate(privateAddress);
+
+        assertEquals(publicAddress.getHost(), actual.getHost());
+        assertEquals(privateAddress.getPort(), actual.getPort());
+    }
+
+    @Test
+    public void testTranslate_whenNotMapped_thenRefresh() {
```
Comment (SaitTalhaNisanci) [inline hazelcast-client/src/test/java/com/hazelcast/client/spi/impl/discovery/HazelcastCloudTranslatorTest.java]: This test looks the same as the above one, I feel like this one does not refresh the lookup table.

File: hazelcast-client/src/main/java/com/hazelcast/client/impl/HazelcastClientInstanceImpl.java [hunk 7]
```diff
@@ -289,18 +304,113 @@ private MetricsRegistryImpl initMetricsRegistry() {
...
+        ClientCloudConfig cloudConfig = networkConfig.getCloudConfig();
+
+        List<String> addresses = networkConfig.getAddresses();
+        boolean addressListProvided = addresses.size() != 0;
+        boolean awsDiscoveryEnabled = awsConfig != null && awsConfig.isEnabled();
+        boolean discoverySpiEnabled = discoveryService != null;
+        String cloudDiscoveryToken = properties.getString(HAZELCAST_CLOUD_DISCOVERY_TOKEN);
+        if (cloudDiscoveryToken != null && cloudConfig.isEnabled()) {
+            throw new IllegalStateException("Ambiguous hazelcast.cloud configuration. "
+                    + "Both property and configuration provided together. Use only one.");
```
Comment (ihsandemir) [inline hazelcast-client/src/main/java/com/hazelcast/client/impl/HazelcastClientInstanceImpl.java]: better to rephrase as: "Both property based and client configuration based settings are provided for Hazelcast cloud discovery together. Use only one."

File: hazelcast-client/src/test/java/com/hazelcast/client/spi/impl/discovery/HazelcastCloudTranslatorTest.java [hunk 8]
```diff
@@ -0,0 +1,112 @@
...
+    @Test
+    public void testRefresh_whenException_thenLogWarning() {
+        HazelcastCloudDiscovery cloudDiscovery = mock(HazelcastCloudDiscovery.class);
+        when(cloudDiscovery.discoverNodes()).thenReturn(lookup);
+
+        translator = new HazelcastCloudAddressTranslator(cloudDiscovery, loggingService);
+
+        translator.refresh();
+    }
+}
```
Comment (Donnerbart) [inline hazelcast-client/src/test/java/com/hazelcast/client/spi/impl/discovery/HazelcastCloudTranslatorTest.java]: Missing newline at EOF

File: hazelcast-client/src/main/java/com/hazelcast/client/spi/impl/discovery/HazelcastCloudAddressTranslator.java [hunk 9]
```diff
@@ -0,0 +1,78 @@
...
+            return null;
+        }
+        Address publicAddress = privateToPublic.get(address);
+        if (publicAddress != null) {
+            return publicAddress;
+        }
+
+        refresh();
+
+        publicAddress = privateToPublic.get(address);
```
Comment (furkansenharputlu) [inline hazelcast-client/src/main/java/com/hazelcast/client/spi/impl/discovery/HazelcastCloudAddressTranslator.java]: `return privateToPublic.get(address);`

**Pull request merged**